### ISSUE TITLE
WPF Rendering issue on any machine

### ISSUE THREAD

**Issue by Erapchu** [dotnet/wpf#4276]
* .NET Version: 5
* Windows version: Windows 10 20H2 (OS Build 19042.867)
* Does the bug reproduce also in WPF for .NET Framework 4.8?: Yes, also .NET Framework 4.5.2, and all .NET Core versions.

 
 
 **Problem description:**
WPF windows stop rendering at all when use BitmapCache on any window's control or root control and first shown WPF window was hidden by set Visibility to Hidden or call Hide() method.
 
 **Actual behavior:** 
There's no exceptions or fatal crashes. All WPF windows that use BitmapCache just stop rendering at all! While FIRST hidden window will not be shown.

 **Expected behavior:**
WPF Windows should render normally.
 
 **Minimal repro:**
Please follow this steps to catch problem on your machine or build any project below. Please, help me resolve it.

https://github.com/Erapchu/GhostWindows
https://github.com/Erapchu/GhostWindowsCore

1. Create empty WPF project.
2. Use property CacheMode="BitmapCache" in MainWindow's Grid;
3. Show MainWindow on start application, then hide it. Show new MainWindow. You can use a code below. Additionally remove StartupUri in App.xaml
4. Press Ctrl + Alt + Del or Win + L and try to interact with showed window (resize, minimize) it should be broken now.

In App.xaml.cs OnStartup overrided method:
`var window1 = new MainWindow();
window1.Show();
await Task.Delay(1000);
window1.Hide();
var window2 = new MainWindow();
window2.Show();`

**Comment by lindexi** [comment]
Duplicate of #2158 ?

**Comment by Erapchu** [comment]
@lindexi Looks like yes. Disable HW acceleration or don't use BitmapCache at all - is not a good idea. Issue #2158 is quite old. I faced it again and temporary solution is disable BitmapCache but performance is bad.

**Comment by ryalanms** [comment]
@Erapchu:  Thank you for the report and investigation.  Are you okay with us closing this as a duplicate?  Thanks.

**Comment by Erapchu** [comment]
Of course. But I want to know why it's happen and how to resolve for now. Future - update will be applied to new .NET version? Not for old like .NET framework?

**Comment by predavid** [comment]
@Erapchu  - it will be applied to the new .NET version.  I will be serviced down to .NET Framework based on community intertest/upvoting. And we would like to track these community responses via a single bug/Issue. Let us use this new one and close #2158

**Comment by Erapchu** [comment]
I did some investigations.
If OS bring up any full-screen window (like UAC prompt when application required administrator or Ctrl + Alt + Del or sign-out), I faced that issue again.
Why BitmapCache broke all windows and how to detect it? Can I subscribe WPF window to WndProc and listen for messages to catch when need to restart rendering (one more question how to do it?).
We just using BitmapCache to improve rendering animations. Can we use something different, because this element invokes problems.

**Comment by Erapchu** [comment]
Setting up BitmapCache to root Grid of the window and Ctrl + Alt + Del invokes problem even if first window was showed, but not focused.
When rendering suspended, WM_PAINT received by window in infinite loop. When remove BitmapCache, problem is vanished, until set it again. When hover mouse over very first window, ALL OTHER windows with BitmapCache rendering normally.
Can anyone deep dive in that problem? @predavid @ryalanms 
develop branch was updated here - https://github.com/Erapchu/GhostWindows

[Ghost windows video, behavior](https://drive.google.com/file/d/1DRvEzbAegLEwRN3u9Mwn29iOitSAjJC6/view?usp=sharing)

**Comment by Erapchu** [comment]
@HyksosSowo the only one solution is don't hide windows, but close and re-open them. Close it completely when bitmapcache is in visual tree 
Anyway we need to sacrifice something

**Comment by EZ64cool** [comment]
Hey, we've been running into this issue for a while as well and have finally come to a stable (but not ideal) solution to this issue.
When creating a window make sure to set `AllowsTransparency` to true (This will also require setting `WindowStyle` to None meaning a custom style and Window class will be needed to add back the default windows functionality).

I couldn't tell you why this works unfortunately (but hopefully it can help narrow down the issue).
I've tried setting `AllowsTransparency` on only windows using `BitmapCache` but it will still cause the problem unless all windows use it.

It's also worth noting that if you're creating `WS_CHILD` window's you'll also need to set the minimum supported OS to windows 8 or higher inside the app.manifest file

```
<compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1">
  <application>
    <!-- Windows 8 -->
    <supportedOS Id="{4a2f28e3-53b9-4441-ba9c-d69d4a4a6e38}" />
  </application>
</compatibility>
``` 

I hope this helps someone out there, as it was a huge pain to find!

**Comment by Valkirie** [comment]
Can still reproduce in .NET8.

**Comment by HyksosSowo** [comment]
> @HyksosSowo the only one solution is don't hide windows, but close and re-open them. Close it completely when bitmapcache is in visual tree Anyway we need to sacrifice something

@Erapchu  we can't close the window. We also can't use the transparency trick as our WPF window is inside a HwndSource.

The GPU virus is still the only solution that reliably work.

**Comment by Erapchu** [comment]
@HyksosSowo try to disable hardware acceleration for that window

https://stackoverflow.com/questions/2169600/how-hardware-acceleration-can-be-disabled-in-wpf

**Comment by Valkirie** [comment]
> @HyksosSowo try to disable hardware acceleration for that window
> 
> https://stackoverflow.com/questions/2169600/how-hardware-acceleration-can-be-disabled-in-wpf

Obviously it works but it's like turning off a few cylinder of your engine because it misbehaves. It suxx...

**Comment by HyksosSowo** [comment]
We are not using WPF for simple dialog box, we are using it to load huge animated graphics, we can't use software rendering... and the funny thing is to get good performance we need HW accelerating + bitmap cache or else it suxx.

**Comment by CycloneRing** [comment]
I also faced same issue, I am using many HLSL shaders and D3DImage in my WPF UI.

Only solution I came up with is to create a empty D3DImage and assign IsFrontBufferAvailableChanged event.

```csharp
internal void IsFrontBufferAvailableChanged_Event (object sender, DependencyPropertyChangedEventArgs e)
{
    if (D3DImageSource.IsFrontBufferAvailable)
    {
        System.Windows.Media.RenderOptions.ProcessRenderMode = RenderMode.Default;
    }
    else
    {
        System.Windows.Media.RenderOptions.ProcessRenderMode = RenderMode.SoftwareOnly;
    }
}
```

Keep this D3DImage somewhere hidden in your window, This works fine and will invalidate BitmapCache objects but the side effect is your UI will flick for ~2 seconds before recovering from WM_PAINT flood.

**Comment by lindexi** [comment]
@CycloneRing Sorry, I don't think you have the same issues. But I think you have this issues: https://github.com/dotnet/wpf/issues/2158
